asm: Allow multiple template string arguments; interpret them as newline-separated

joshtriplett opened the pull request:
Allow the `asm!` macro to accept a series of template arguments, and interpret them as if they were concatenated with a '\n' between them. This allows writing an `asm!` where each line of assembly appears in a separate template string argument.

This syntax makes it possible for rustfmt to reliably format and indent each line of assembly, without risking changes to the inside of a template string. It also avoids the complexity of having the user carefully format and indent a multi-line string (including where to put the surrounding quotes), and avoids the extra indentation and lines of a call to `concat!`.

For example, rewriting the second example from the [blog post on the new inline assembly syntax](https://blog.rust-lang.org/inside-rust/2020/06/08/new-inline-asm.html) using multiple template strings:

```rust

fn main() {
    let mut bits = [0u8; 64];
    for value in 0..=1024u64 {
        let popcnt;
        unsafe {
            asm!(
                "    popcnt {popcnt}, {v}",
                "2:",
                "    blsi rax, {v}",
                "    jz 1f",
                "    xor {v}, rax",
                "    tzcnt rax, rax",
                "    stosb",
                "    jmp 2b",
                "1:",
                v = inout(reg) value => _,
                popcnt = out(reg) popcnt,
                out("rax") _, // scratch
                inout("rdi") bits.as_mut_ptr() => _,
            );
        }
        println!("bits of {}: {:?}", value, &bits[0..popcnt]);
    }
}
```

Note that all the template strings must appear before all other arguments; you cannot, for instance, provide a series of template strings intermixed with the corresponding operands.

joshtriplett committed:
asm: Unify pseudo-keyword parsing using `eat`, rather than a final `expect`

Currently, `asm!` parsing uses an `expect` for the last parsed
pseudo-keyword (`sym`), which makes it difficult to extend without
simultaneously refactoring. Use `eat` for the last pseudo-keyword, and
then add an `else` that fails parsing. No change to error output.

joshtriplett committed:
asm: When pretty-printing, don't escape characters twice

pprust uses `print_string` to write out the template string, and
`print_string` already calls `escape_debug`, so `impl fmt::Display for
InlineAsmTemplatePiece` shouldn't do an additional `escape_debug`.

This fixes a pretty-printing bug that translated
`asm!("...\n...")`
to
`asm!("...\\n...")`

joshtriplett committed:
asm: Allow multiple template strings; interpret them as newline-separated

Allow the `asm!` macro to accept a series of template arguments, and
interpret them as if they were concatenated with a '\n' between them.
This allows writing an `asm!` where each line of assembly appears in a
separate template string argument.

This syntax makes it possible for rustfmt to reliably format and indent
each line of assembly, without risking changes to the inside of a
template string. It also avoids the complexity of having the user
carefully format and indent a multi-line string (including where to put
the surrounding quotes), and avoids the extra indentation and lines of a
call to `concat!`.

For example, rewriting the second example from the [blog post on the new
inline assembly
syntax](https://blog.rust-lang.org/inside-rust/2020/06/08/new-inline-asm.html)
using multiple template strings:

```rust

fn main() {
    let mut bits = [0u8; 64];
    for value in 0..=1024u64 {
        let popcnt;
        unsafe {
            asm!(
                "    popcnt {popcnt}, {v}",
                "2:",
                "    blsi rax, {v}",
                "    jz 1f",
                "    xor {v}, rax",
                "    tzcnt rax, rax",
                "    stosb",
                "    jmp 2b",
                "1:",
                v = inout(reg) value => _,
                popcnt = out(reg) popcnt,
                out("rax") _, // scratch
                inout("rdi") bits.as_mut_ptr() => _,
            );
        }
        println!("bits of {}: {:?}", value, &bits[0..popcnt]);
    }
}
```

Note that all the template strings must appear before all other
arguments; you cannot, for instance, provide a series of template
strings intermixed with the corresponding operands.

In order to get srcloc mappings right for macros that generate
multi-line string literals, create one line_span for each
line in the string literal, each pointing to the macro.

Make `rustc_parse_format::Parser::curarg` `pub`, so that we can
propagate it from one template string argument to the next.

rust-highfive commented:
r? @eddyb

(rust_highfive has picked a reviewer for you, use r? to override)

joshtriplett commented:
r? @Amanieu

Amanieu commented on src/librustc_ast/ast.rs:
```diff
@@ -1904,6 +1904,7 @@ bitflags::bitflags! {
 pub enum InlineAsmTemplatePiece {
     String(String),
     Placeholder { operand_idx: usize, modifier: Option<char>, span: Span },
+    TemplateBoundary,
```
Isn't this just a `String("\n")`? It would be nice not to have to deal with `TemplateBoundary` in the later stages of the compiler.

joshtriplett replied:
Having a separate `TemplateBoundary` allows transforming inline assembly in the AST back to the exact set of template string arguments originally provided; I'd prefer to retain that property.

Amanieu replied:
We already don't preserve the template strings exactly since `{}` and `{name}` placeholders are turned into `{0}` in the AST. I don't think there is much value in preserving split template strings.

joshtriplett replied:
@Amanieu Switching to a different approach that doesn't use `TemplateBoundary`, as discussed on Zulip.

Amanieu commented on src/librustc_builtin_macros/asm.rs:
```diff
@@ -330,155 +349,168 @@ fn parse_reg<'a>(
         }
+
+        if parser.line_spans.is_empty() {
+            line_spans.push(template_sp);
```
This is wrong if this component of the template string contains multiple lines, but comes from a macro (like `concat!`) so we don't have line spans for it. You need to modify the parsers to always generate enough line spans (pointing to `template_sp`) for the number of lines in the template component.

Amanieu replied:
Also make sure that trailing newlines are properly handled since a line span may not be generated for them.

joshtriplett replied:
Fixed, and I've added several test cases to cover that.

Amanieu commented on src/librustc_builtin_macros/asm.rs:
```diff
@@ -147,8 +147,27 @@ fn parse_args<'a>(
+            // things it could have been.
+            match template.kind {
+                ast::ExprKind::Lit(ast::Lit { kind: ast::LitKind::Str(..), .. }) => {}
+                ast::ExprKind::MacCall(..) => {}
```
I don't think `MacCall` is ever reachable here since it would have been expanded already, but I'm not 100% sure.

joshtriplett replied:
If that were true, we wouldn't need the call to `expr_to_spanned_string` in `expand_preparsed_asm` (which hasn't been called yet at this point). Removing `MacCall` from this match causes `concat!` or other macros to fail if passed in the second or subsequent template argument of an `asm!`. I have a test case for this.

joshtriplett commented:
Fixing clippy...

Amanieu reviewed (commented):
Have you checked that this is correctly handling implicitly number arguments in the template over multiple template strings? For example `asm!("{}", "{}", in(reg) a, in(reg) b)`.

joshtriplett commented:
@Amanieu
> Have you checked that this is correctly handling implicitly number arguments in the template over multiple template strings? For example `asm!("{}", "{}", in(reg) a, in(reg) b)`.

I've now added a couple more tests specifically for that case, both using implicit numbering and using explicit numbering (including in reverse).

Working on making those tests pass, now; good catch.

rust-highfive commented:
The job `mingw-check` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=31994&view=logs&jobId=74054d28-c774-5fab-ecc2-352fe71b230e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/31994/logs/65), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/31994/logs/65)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[section]Starting: Linux mingw-check
##[section]Starting: Initialize job
Agent name: 'Hosted Agent'
Agent machine name: 'fv-az578'
Current agent version: '2.170.1'
##[group]Operating System
16.04.6
LTS
LTS
##[endgroup]
##[group]Virtual Environment
Environment: ubuntu-16.04
Version: 20200604.1
Included Software: https://github.com/actions/virtual-environments/blob/ubuntu16/20200604.1/images/linux/Ubuntu1604-README.md
##[endgroup]
Agent running as: 'vsts'
Prepare build directory.
Set build variables.
Download all required tasks.
Download all required tasks.
Downloading task: Bash (3.163.3)
Checking job knob settings.
   Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
   Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
Start tracking orphan processes.
##[section]Finishing: Initialize job
##[section]Starting: Configure Job Name
==============================================================================
---
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/355a7d89-603e-402a-9cdb-76f22ac1c7f3.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/73364/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/73364/merge:refs/remotes/pull/73364/merge
---
 ---> f883e675ad62
Step 6/7 : ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
 ---> Using cache
 ---> c0b156eb069c
Step 7/7 : ENV SCRIPT python3 ../x.py test src/tools/expand-yaml-anchors &&            python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu &&            python3 ../x.py build --stage 0 src/tools/build-manifest &&            python3 ../x.py test --stage 0 src/tools/compiletest &&            python3 ../x.py test src/tools/tidy &&            python3 ../x.py doc --stage 0 src/libstd &&            /scripts/validate-toolstate.sh
 ---> 8541bab6b38c
Successfully built 8541bab6b38c
Successfully tagged rust-ci:latest
Built container sha256:8541bab6b38c07f1b7eb787539b9cbe93daa6ac4458d3d7bd8a8921622a14ba1
---
    Checking rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
    Checking chalk-rust-ir v0.10.0
    Checking rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
    Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
    Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
    Checking chalk-solve v0.10.0
    Checking rustc_parse v0.0.0 (/checkout/src/librustc_parse)
    Checking rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
    Checking rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
    Checking toml v0.5.3
    Checking url v2.1.0
    Checking cargo_metadata v0.9.1
    Checking clippy_lints v0.0.212 (/checkout/src/tools/clippy/clippy_lints)
error[E0004]: non-exhaustive patterns: `&TemplateBoundary` not covered
     |
474  |                     match piece {
474  |                     match piece {
     |                           ^^^^^ pattern `&TemplateBoundary` not covered
    ::: /checkout/src/librustc_ast/ast.rs:1907:5
     |
1907 |     TemplateBoundary,
     |     ---------------- not covered
     |     ---------------- not covered
     |
     = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
     = note: the matched value is of type `&rustc_ast::ast::InlineAsmTemplatePiece`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0004`.
error: could not compile `clippy_lints`.
---
  local time: Mon Jun 15 07:13:27 UTC 2020
  network time: Mon, 15 Jun 2020 07:13:27 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/73364/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/73364/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (3352) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Amanieu commented on src/librustc_builtin_macros/asm.rs:
```diff
@@ -330,155 +349,180 @@ fn parse_reg<'a>(
+        };
+
+        let template_str = &template_str.as_str();
+        let template_num_lines = 1 + template_str.matches('\n').count();
```
Move this down into the `if` so we only calculate this if needed.

Amanieu commented on src/librustc_builtin_macros/asm.rs:
```diff
@@ -330,155 +349,180 @@ fn parse_reg<'a>(
         }
+
+        if parser.line_spans.is_empty() {
+            line_spans.extend([template_sp].iter().copied().cycle().take(template_num_lines));
```
Use `iter::repeat(template_sp).take(template_num_lines)`.

Amanieu commented on src/librustc_parse_format/lib.rs:
```diff
@@ -243,11 +243,14 @@ impl<'a> Iterator for Parser<'a> {
-                self.line_spans.push(start.to(end));
-                self.cur_line_start = self.input.len();
+                let span = start.to(end);
+                if self.line_spans.last().map(|last_span| last_span.end) != Some(span.end) {
+                    self.line_spans.push(span);
+                    self.cur_line_start = self.input.len();
+                }
```
```suggestion
                if self.line_spans.last() != Some(span) {
                    self.line_spans.push(span);
                }
```

joshtriplett replied:
Done; confirmed that it seems to pass tests. I was concerned that this would cause a problem for other code using `cur_line_start`, but it looks like none of the other code using `cur_line_start` can run once this condition has happened.

Amanieu commented on src/librustc_builtin_macros/asm.rs:
```diff
@@ -147,8 +147,27 @@ fn parse_args<'a>(
+                ast::ExprKind::Lit(ast::Lit { kind: ast::LitKind::Str(..), .. }) => {}
+                ast::ExprKind::MacCall(..) => {}
+                _ => {
+                    let errstr = "expected operand, options, or additional template string";
```
I'm not too happy about this error message since unlike the one from `expect` it doesn't give you a list of all the available operand types for asm (`in`, `out`, `sym`, etc). However I don't think we can do much better here anyways, so it's probably fine as it is.

joshtriplett replied:
I originally did copy the full list into this error string; I just changed it because it seemed unwieldy. I'm happy to put it back if you'd prefer. It would look like this:

```
error: expected `const`, `in`, `inlateout`, `inout`, `lateout`, `options`, `out`, `sym`, or additional template string
```

Or perhaps this:

```
error: expected additional template string, operand (`const`, `in`, `inlateout`, `inout`, `lateout`, `out`, `sym`), or `options`
```

Does that seem better?

Amanieu replied:
I think it's fine as it is, I can't really think of anything better than what you already have.

joshtriplett replied:
Alright, I'll leave it as-is.

joshtriplett commented on src/librustc_parse_format/lib.rs:
```diff
@@ -243,7 +243,7 @@ impl<'a> Iterator for Parser<'a> {
+            if self.is_literal {
                 let start = self.to_span_index(self.cur_line_start);
                 let end = self.to_span_index(self.input.len());
                 self.line_spans.push(start.to(end));
```
Already done, concurrently with your comment.

Amanieu commented:
Can you update src/doc/unstable-book with the same changes you made to the RFC?

petrochenkov commented:
>and interpret them as if they were concatenated with a '\n' between them

So it's not just C-style concatenation of neighboring string literals for people accustomed to it.
This certainly makes it more useful.

joshtriplett committed:
asm: Update chapter in unstable book for multiple template string arguments

Update all examples to use the new formatting, and update explanations
to document it.

joshtriplett commented:
@Amanieu 
> Can you update `src/doc/unstable-book/src/library-features/asm.md` with the same changes you made to the RFC?

Done; thanks!

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit fd9ed30e4b19457cc7059d1b08832edd6d2cbfb8 has been approved by `Amanieu`

<!-- @bors r=Amanieu fd9ed30e4b19457cc7059d1b08832edd6d2cbfb8 -->
<!-- homu: {"type":"Approved","sha":"fd9ed30e4b19457cc7059d1b08832edd6d2cbfb8","approver":"Amanieu"} -->
